servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererObserverTest.java

```diff
@@ -107,9 +113,37 @@ void srvQueryTriggersNewDiscoveryObserver() throws Exception {
     private void testNewDiscoveryObserver(BiFunction<DnsClient, String, Publisher<?>> publisherFactory,
                                           String expectedName) throws Exception {
         BlockingQueue<String> newDiscoveryCalls = new LinkedBlockingDeque<>();
-        DnsClient client = dnsClient(name -> {
-            newDiscoveryCalls.add(name);
-            return NoopDnsDiscoveryObserver.INSTANCE;
+        final AtomicBoolean discoveryCanceledCalled = new AtomicBoolean();
+        DnsClient client = dnsClient(new DnsServiceDiscovererObserver() {
+            @Override
+            public DnsDiscoveryObserver onNewDiscovery(final String name) {
+                fail("This method must not be called anymore when overridden");
```

Comment:
When ST plugs in any observer, we always try-catch its callbacks to make sure misbehaving observability does not impact the main execution flow. This assertion will not fail the test. Instead, consider using a `BlockingQueue` where you can add `AssertionException` and they verify that the queue is empty at the end of the test. We have a utility for that, see `TestUtils.assertNoAsyncErrors(...)` and how it's used by other tests.